### ISSUE TITLE
Get rid of most/all uses of java.util.Properties

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#6391]
The next time that we plan to make an incompatible change to the API, we should replace as many uses of `java.util.Properties` as possible. When the original implementation was done, it was `Properties`, backed by `Hashtable`, which was basically the only option, but it is doesn't allow `null` values and its always synchronized (ie slow), two major drawbacks. Since Java 1.2, there have been better options available like `java.util.HashMap`.

### Proposed solution

- [ ] Bindings (ie variables) replace with Map<String, Object>
- [ ] URL query parameters - replace with Map<String, String[]> (and get rid of homegrown parsing methods)
- [ ] Importers - as a side effect of URL parameters
- [ ] com.google.refine.exporters.Exporter.export() - replace with Map<String, String>
- [ ] writer options - Change, HistoryEntry, Cell, anything serialized as JSON - replace with Map<String, String> (or something more specific?

### Alternatives considered

We could leave the API as it is, but a more modern interface would benefit both core developers and extension writers.

### Additional context

The initial motivation for this is being able to have `null` values in bindings, but there's really almost no place where using the old class is needed or a good idea.

**Comment by thadguidry** [comment]
Would Java Record not be more appropriate for a few of those bullet points especially where immutability and validity without synchronization is desirable?

**Comment by tfmorris** [comment]
I don't think so, but you don't say enough about why you, apparently, do for me to be able to weight the evidence. Which cases did you think it would be better for and why?

**Comment by thadguidry** [comment]
Importers, exporters, and writer options.

I thought that being able to easily stringify things could help in error messaging as well as the reproducibility project. As an example, Template options could be easier to print and share, etc. The other benefit would be less boilerplate which helps contributors.

**Comment by tfmorris** [comment]
The maps used by importers and exporters are part of the extension points, so those need to be open ended. If writer options are a fixed (not sure off the top of my head), it may be possible to use a locked down implementation like Record there.

I've updated the issue with more context about the current usage and effective contract and will add links to individual issues for each migration as they are created.

**Comment by EliasStihl** [comment]
Hello, @Karindev, @Ishaqezas,@Edvinnordling, and I have one week to address an open-source issue as part of our coursework at KTH. We would like to contribute to this particular task. Could you please consider assigning us as the assignees for the issue?

**Comment by tfmorris** [comment]
Hello @EliasStihl, @karindev, @Ishaqezaz,and @Edvinnordling. Thank you for volunteering! Unfortunately, I think this is probably a bit too big of a task for a week, even with four of you. It's actually an epic with multiple sub-issues and probably a bit too complex for developers who are new to the project.

We have a "Good First Issue" tag which can be helpful in finding issues that might be appropriate (although they could be too small for four people). Let us know if you need help in identifying something appropriate to work on.

**Comment by EliasStihl** [comment]
Thank you for your response!

There are three main requirements in our task:
1. The issue should be somewhat related to refactoring.
2. The issue is non-trivial in the sense that it does not affect only single, local functions, but at least multiple functions, units, classes, files, or modules.
3. The issue should be open and have no assignee

If you could help us find such an issue (or possibly create one), that would be awesome!

**Comment by tfmorris** [comment]
@EliasStihl I wrote up #6403, which is a piece of this epic, as a possible candidate for your group to consider, unless @wetneb has a better suggestion. It will affect about 10 modules and require addressing public API compatibility.

**Comment by EliasStihl** [comment]
Great! Thank you very much, we will try to solve that one!

**Comment by tfmorris** [comment]
@Edvinnordling @karindev I'm not trying to ignore you. When I looked at the Exporters side of things again, I was no longer convinced that the benefit was worth the API churn, but let me take another look. Unfortunately, I don't have any other refactoring tasks groomed and ready to go, unless perhaps @wetneb has something in his back pocket.

**Comment by tfmorris** [comment]
@karindev, @EliasStihl,  @Ishaqezaz, @Edvinnordling - I've written up the exporter refactoring task (see link above). If you're up for a bit of an experimental challenge, please feel free to tackle it.

**Comment by karindev** [comment]
Thank you so much, we'll take a look at it!

**Comment by tfmorris** [comment]
> Would Java Record not be more appropriate for a few of those bullet points especially where immutability and validity without synchronization is desirable?

Java Records would be a candidate for WriterOptions in #7523, but they require Java 16, meaning that use of them is blocked on #7281 .
